File: scripts/six_degrees_test.js
#!/usr/bin/env bun
/**
 * 六度分隔理论测试脚本 (JavaScript)
 * 读取 dist/all.json，对所有 C(n,2) 节点对计算最短路径，
 * 统计最大距离是否超过 6 度。
 *
 * 用法: bun scripts/six_degrees_test.js
 */

import { readFileSync } from "fs";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const ALL_JSON = resolve(__dirname, "..", "dist", "all.json");

/** BFS 计算 start 到所有其他节点的最短距离（边数），不可达为 -1 */
function bfsDistance(adj, start) {
  const n = adj.length;
  const dist = new Int32Array(n).fill(-1);
  dist[start] = 0;
  const q = [start];
  let head = 0;
  while (head < q.length) {
    const u = q[head++];
    for (const v of adj[u]) {
      if (dist[v] === -1) {
        dist[v] = dist[u] + 1;
        q.push(v);
      }
    }
  }
  return dist;
}

/** 构建无向图邻接表，返回 { urlToIdx, allUrls, adj } */
function buildGraph(sites) {
  const urlSet = new Set();
  for (const site of sites) {
    urlSet.add(site.url);
    for (const friend of site.friends || []) {
      urlSet.add(friend.url);
    }
  }

  const allUrls = [...urlSet].sort();
  const urlToIdx = new Map();
  allUrls.forEach((u, i) => urlToIdx.set(u, i));

  const n = allUrls.length;
  const adj = Array.from({ length: n }, () => []);
  for (const site of sites) {
    const src = urlToIdx.get(site.url);
    for (const friend of site.friends || []) {
      const dst = urlToIdx.get(friend.url);
      adj[src].push(dst);
      adj[dst].push(src);
    }
  }

  return { urlToIdx, allUrls, adj };
}

function main() {
  console.error(`读取 ${ALL_JSON} ...`);
  const raw = readFileSync(ALL_JSON, "utf-8");
  const data = JSON.parse(raw);

  const sites = data.sites;
  console.error(`核心节点: ${sites.length}`);

  const { allUrls, adj } = buildGraph(sites);
  const n = allUrls.length;
  const edgeCount = adj.reduce((sum, a) => sum + a.length, 0) / 2;
  console.error(`总节点数: ${n}`);
  console.error(`总边数: ${edgeCount}`);

  // 找连通分量
  const component = new Int32Array(n).fill(-1);
  const compNodesMap = new Map(); // compId -> [nodeIdx...]
  let compId = 0;

  for (let i = 0; i < n; i++) {
    if (component[i] !== -1) continue;
    const q = [i];
    component[i] = compId;
    const nodes = [];
    let head = 0;
    while (head < q.length) {
      const u = q[head++];
      nodes.push(u);
      for (const v of adj[u]) {
        if (component[v] === -1) {
          component[v] = compId;
          q.push(v);
        }
      }
    }
    compNodesMap.set(compId, nodes);
    compId++;
  }

  console.error(`连通分量数: ${compId}`);

  // 按大小排序，优先大分量
  const sortedComps = [...compNodesMap.entries()].sort(
    (a, b) => b[1].length - a[1].length,
  );

  let maxDist = 0;
  let maxPair = null;
  const distCount = {};
  let unreachable = 0;
  let totalPairs = 0;
  let processedNodes = 0;

  const startTime = Date.now();

  for (const [cid, compNodes] of sortedComps) {
    const size = compNodes.length;
    for (let aIdx = 0; aIdx < size; aIdx++) {
      if (aIdx % 50 === 0 || aIdx === size - 1) {
        const progress = ((processedNodes + aIdx + 1) / n) * 100;
        const elapsed = ((Date.now() - startTime) / 1000).toFixed(1);
        process.stderr.write(
          `\r  进度: ${progress.toFixed(1)}%  (分量 ${cid + 1}/${compId}, 大小 ${size}, 节点 ${aIdx + 1}/${size})  ${elapsed}s`,
        );
      }

      const a = compNodes[aIdx];
      const dist = bfsDistance(adj, a);

      for (let bIdx = aIdx + 1; bIdx < size; bIdx++) {
        const b = compNodes[bIdx];
        const d = dist[b];
        if (d === -1) {
          unreachable++;
        } else {
          totalPairs++;
          distCount[d] = (distCount[d] || 0) + 1;
          if (d > maxDist) {
            maxDist = d;
            maxPair = [a, b];
          }
        }
      }
    }
    processedNodes += size;
  }

  process.stderr.write("\n\n");

  console.log(`\n距离分布:`);
  const maxCount = Math.max(...Object.values(distCount));
  for (const d of Object.keys(distCount)
    .map(Number)
    .sort((a, b) => a - b)) {
    const bar = "█".repeat(
      Math.max(1, Math.round((distCount[d] * 60) / maxCount)),
    );
    console.log(
      `  ${String(d).padStart(2)} 度: ${String(distCount[d]).padStart(8)}  ${bar}`,
    );
  }

  console.log(`\n最大距离: ${maxDist} 度`);
  if (maxPair) {
    // 回溯路径
    const [aIdx, bIdx] = maxPair;
    const distA = bfsDistance(adj, aIdx);
    const path = [bIdx];
    let cur = bIdx;
    while (cur !== aIdx) {
      for (const prev of adj[cur]) {
        if (distA[prev] === distA[cur] - 1) {
          path.push(prev);
          cur = prev;
          break;
        }
      }
    }
    path.reverse();
    console.log(`\n最大距离点对:`);
    console.log(`  起点: ${allUrls[path[0]]}`);
    console.log(`  终点: ${allUrls[path[path.length - 1]]}`);
    console.log(`  路径 (${path.length} 个节点, ${path.length - 1} 条边):`);
    for (let i = 0; i < path.length; i++) {
      const url = allUrls[path[i]];
      const site = sites.find((s) => s.url === url);
      const name = site ? site.name : url;
      console.log(`    ${i}: ${name} (${url})`);
    }
  }

  const beyondSix = Object.entries(distCount)
    .filter(([d]) => Number(d) > 6)
    .reduce((sum, [, v]) => sum + v, 0);

  console.log(`\n${"=".repeat(50)}`);
  console.log(`六度分隔理论检验:`);
  console.log(`  超过 6 度的点对数: ${beyondSix}`);
  console.log(`  总可达点对数: ${totalPairs}`);
  if (totalPairs > 0) {
    const pct = (beyondSix / totalPairs) * 100;
    console.log(`  占比: ${pct.toFixed(4)}%`);
    if (beyondSix === 0) {
      console.log(`  ✅ 所有可达点对都在 6 度以内，符合六度分隔理论！`);
    } else {
      console.log(`  ❌ 有 ${beyondSix} 对超过 6 度，不符合六度分隔理论`);
    }
  }
  console.log(`${"=".repeat(50)}`);

  const elapsed = ((Date.now() - startTime) / 1000).toFixed(1);
  console.log(`\n总耗时: ${elapsed}s`);
}

main();

File: scripts/six_degrees_test.py
#!/usr/bin/env python3
"""
六度分隔理论测试脚本
读取 dist/all.json，对所有 C(n,2) 节点对计算最短路径，
统计最大距离是否超过 6 度。
"""

import json
import sys
from collections import deque
from pathlib import Path

ALL_JSON = Path(__file__).resolve().parent.parent / "dist" / "all.json"


def build_graph(sites: list[dict]) -> tuple[dict[str, int], list[str], dict[int, set[int]]]:
    """构建无向图邻接表，返回 (url→idx, idx→url, adj)"""
    # 收集所有 URL
    url_set: set[str] = set()
    for site in sites:
        url_set.add(site["url"])
        for friend in site.get("friends", []):
            url_set.add(friend["url"])

    all_urls = sorted(url_set)
    url_to_idx = {u: i for i, u in enumerate(all_urls)}

    adj: dict[int, set[int]] = {i: set() for i in range(len(all_urls))}

    for site in sites:
        src = url_to_idx[site["url"]]
        for friend in site.get("friends", []):
            dst = url_to_idx[friend["url"]]
            adj[src].add(dst)
            adj[dst].add(src)

    return url_to_idx, all_urls, adj


def bfs_distance(adj: dict[int, set[int]], start: int) -> list[int]:
    """BFS 计算 start 到所有其他节点的最短距离（边数），不可达为 -1"""
    n = len(adj)
    dist = [-1] * n
    dist[start] = 0
    q = deque([start])
    while q:
        u = q.popleft()
        for v in adj[u]:
            if dist[v] == -1:
                dist[v] = dist[u] + 1
                q.append(v)
    return dist


def main():
    print(f"读取 {ALL_JSON} ...")
    with open(ALL_JSON, "r") as f:
        data = json.load(f)

    sites = data["sites"]
    print(f"核心节点: {len(sites)}")

    url_to_idx, all_urls, adj = build_graph(sites)
    n = len(all_urls)
    print(f"总节点数: {n}")
    print(f"总边数: {sum(len(v) for v in adj.values()) // 2}")

    max_dist = 0
    max_pair: tuple[int, int] | None = None
    dist_count: dict[int, int] = {}  # 距离 → 出现次数
    unreachable = 0
    total_pairs = 0

    # 对每个连通分量做 BFS
    # 只在同一个连通分量内比较
    component: list[int] = [-1] * n
    comp_id = 0
    comp_sizes: list[int] = []

    for i in range(n):
        if component[i] != -1:
            continue
        # BFS 标记连通分量
        q = deque([i])
        component[i] = comp_id
        comp_nodes: list[int] = []
        while q:
            u = q.popleft()
            comp_nodes.append(u)
            for v in adj[u]:
                if component[v] == -1:
                    component[v] = comp_id
                    q.append(v)

        comp_sizes.append(len(comp_nodes))
        comp_id += 1

    # 按分量大小排序，优先处理大分量
    # 重新遍历：按节点所属分量分组
    comp_nodes_map: dict[int, list[int]] = {}
    for i in range(n):
        c = component[i]
        comp_nodes_map.setdefault(c, []).append(i)

    processed_nodes = 0
    for cid in sorted(comp_nodes_map, key=lambda c: len(comp_nodes_map[c]), reverse=True):
        comp_nodes = comp_nodes_map[cid]
        size = len(comp_nodes)
        for a_idx, a in enumerate(comp_nodes):
            if a_idx % 20 == 0 or a_idx == size - 1:
                progress = (processed_nodes + a_idx + 1) / n * 100
                print(
                    f"\r  进度: {progress:.1f}%  (分量 {cid+1}/{comp_id}, 大小 {size}, 节点 {a_idx+1}/{size})",
                    end="",
                    file=sys.stderr,
                    flush=True,
                )
            dist = bfs_distance(adj, a)
            for b in comp_nodes[a_idx + 1 :]:
                d = dist[b]
                if d == -1:
                    unreachable += 1
                else:
                    total_pairs += 1
                    dist_count[d] = dist_count.get(d, 0) + 1
                    if d > max_dist:
                        max_dist = d
                        max_pair = (a, b)
        processed_nodes += size

    print("", file=sys.stderr)  # newline

    print(f"\n连通分量数: {comp_id}")
    print(f"最大分量大小: {max(comp_sizes)}")
    print(f"总可达点对数: {total_pairs}")
    print(f"不可达点对数: {unreachable}")

    print(f"\n距离分布:")
    for d in sorted(dist_count):
        bar = "█" * max(1, dist_count[d] * 60 // max(dist_count.values()))
        print(f"  {d:2d} 度: {dist_count[d]:>8d}  {bar}")

    print(f"\n最大距离: {max_dist} 度")

    if max_pair is not None:
        a_url, b_url = all_urls[max_pair[0]], all_urls[max_pair[1]]
        # 找出实际路径
        dist_a = bfs_distance(adj, max_pair[0])
        path = [max_pair[1]]
        cur = max_pair[1]
        while cur != max_pair[0]:
            for prev in adj[cur]:
                if dist_a[prev] == dist_a[cur] - 1:
                    path.append(prev)
                    cur = prev
                    break
        path.reverse()
        print(f"\n最大距离点对:")
        print(f"  起点: {a_url}")
        print(f"  终点: {b_url}")
        print(f"  路径 ({len(path)} 个节点, {len(path)-1} 条边):")
        for i, idx in enumerate(path):
            site = next((s for s in sites if s["url"] == all_urls[idx]), None)
            name = site["name"] if site else all_urls[idx]
            print(f"    {i}: {name} ({all_urls[idx]})")

    # 六度理论检验
    beyond_six = sum(v for d, v in dist_count.items() if d > 6)
    print(f"\n{'='*50}")
    print(f"六度分隔理论检验:")
    print(f"  超过 6 度的点对数: {beyond_six}")
    print(f"  总可达点对数: {total_pairs}")
    if total_pairs > 0:
        pct = beyond_six / total_pairs * 100
        print(f"  占比: {pct:.4f}%")
        if beyond_six == 0:
            print(f"  ✅ 所有可达点对都在 6 度以内，符合六度分隔理论！")
        else:
            print(f"  ❌ 有 {beyond_six} 对超过 6 度，不符合六度分隔理论")
    print(f"{'='*50}")


if __name__ == "__main__":
    main()
